ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Aux.hs

```diff
@@ -0,0 +1,545 @@
+                                slotNo
+    }
+
+applyEpochTransition :: Gen.Config
+                     -> CC.SlotNumber
+                     -> CC.ChainValidationState
+                     -> CC.ChainValidationState
+applyEpochTransition cfg slotNo state = state {
+      CC.cvsUpdateState = CC.epochTransition
+                            (mkEpochEnvironment cfg state)
+                            (CC.cvsUpdateState state)
+                            slotNo
+    }
+
+validateHeader :: MonadError CC.ChainValidationError m
+               => CC.ValidationMode
+               -> U.Iface.State -> CC.AHeader ByteString -> m ()
+validateHeader validationMode updState hdr =
+    flip runReaderT validationMode $
+      CC.headerIsValid updState hdr
+
+validateBody :: MonadError CC.ChainValidationError m
+             => CC.ValidationMode
+             -> CC.ABlock ByteString
+             -> CC.BodyEnvironment -> CC.BodyState -> m CC.BodyState
+validateBody validationMode block bodyEnv bodyState =
+    flip runReaderT validationMode $
+      CC.updateBody bodyEnv bodyState block
+
+validateBlock :: MonadError CC.ChainValidationError m
+              => Gen.Config
+              -> CC.ValidationMode
+              -> CC.ABlock ByteString
+              -> CC.HeaderHash
+              -> CC.ChainValidationState -> m CC.ChainValidationState
+validateBlock cfg validationMode block blkHash state = do
+    validateHeader validationMode updState (CC.blockHeader block)
+    bodyState' <- validateBody validationMode block bodyEnv bodyState
+    return state {
+          CC.cvsLastSlot        = CC.blockSlot block
+        , CC.cvsPreviousHash    = Right blkHash
+        , CC.cvsUtxo            = CC.utxo            bodyState'
+        , CC.cvsUpdateState     = CC.updateState     bodyState'
+        , CC.cvsDelegationState = CC.delegationState bodyState'
+        }
+  where
+    updState  = CC.cvsUpdateState state
+    bodyEnv   = mkBodyEnv cfg (getProtocolParams state) (CC.blockSlot block)
+    bodyState = mkBodyState state
+
+-- | Apply a boundary block
+--
+-- We don't validate boundary blocks and so this cannot fail.
+--
+-- TODO:
+-- * Should we check that the prev hash lines up?
+-- * Is the cvsLastSlot update compatible with what we do in consensus?
+applyBoundary :: Gen.Config
+              -> CC.ABoundaryBlock a
+              -> CC.HeaderHash
+              -> CC.ChainValidationState -> CC.ChainValidationState
+applyBoundary cfg blk blkHash state = state {
+      CC.cvsPreviousHash = Right blkHash
+    , CC.cvsLastSlot     = CC.SlotNumber $ epochSlots * CC.boundaryEpoch hdr
+    }
+  where
+    hdr = CC.boundaryHeader blk
+    CC.EpochSlots epochSlots = Gen.configEpochSlots cfg
+
+applyScheduledDelegations :: Seq D.Sched.ScheduledDelegation
+                          -> Delegation.Map -> Delegation.Map
+applyScheduledDelegations update (Delegation.Map del) =
+    Delegation.Map $ Foldable.foldl' (flip applyOne) del update
```

Comment:
Is this `foldl'` correct?

Reply:
Added comment:

```
    -- The order in which we apply the updates does not matter, because the spec
    -- says: "Any given key can issue at most one certificate in a given slot."
```